modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/inlineschema/TupleWithSchemaMarshalling.java

```diff
@@ -0,0 +1,333 @@
+
+        for (int i = 0; i < columns.length; i++) {
+            builder.appendString(columns[i]);
+            builder.appendInt(types[i].id());
+        }
+
+        return builder;
+    }
+
+    private static BinaryTupleBuilder valueBuilder(Object[] values, ColumnType[] types) {
+        BinaryTupleBuilder builder = new BinaryTupleBuilder(values.length, values.length);
+
+        for (int i = 0; i < values.length; i++) {
+            ColumnType type = types[i];
+            Object v = values[i];
+
+            appender(type).accept(builder, v);
+        }
+
+        return builder;
+    }
+
+    private static BiConsumer<BinaryTupleBuilder, Object> appender(ColumnType type) {
+        switch (type) {
+            case NULL:
+                return (b, v) -> b.appendNull();
+            case BOOLEAN:
+                return (b, v) -> b.appendBoolean(checkType(Boolean.class, v));
+            case INT8:
+                return (b, v) -> b.appendByte(checkType(Byte.class, v));
+            case INT16:
+                return (b, v) -> b.appendShort(checkType(Short.class, v));
+            case INT32:
+                return (b, v) -> b.appendInt(checkType(Integer.class, v));
+            case INT64:
+                return (b, v) -> b.appendLong(checkType(Long.class, v));
+            case FLOAT:
+                return (b, v) -> b.appendFloat(checkType(Float.class, v));
+            case DOUBLE:
+                return (b, v) -> b.appendDouble(checkType(Double.class, v));
+            case STRING:
+                return (b, v) -> b.appendString(checkType(String.class, v));
+            case DECIMAL:
+                return (b, v) -> b.appendDecimal(checkType(BigDecimal.class, v), ((BigDecimal) v).scale());
+            case DATE:
+                return (b, v) -> b.appendDate(checkType(LocalDate.class, v));
+            case TIME:
+                return (b, v) -> b.appendTime(checkType(LocalTime.class, v));
+            case DATETIME:
+                return (b, v) -> b.appendDateTime(checkType(LocalDateTime.class, v));
+            case TIMESTAMP:
+                return (b, v) -> b.appendTimestamp(checkType(Instant.class, v));
+            case UUID:
+                return (b, v) -> b.appendUuid(checkType(UUID.class, v));
+            case BYTE_ARRAY:
+                return (b, v) -> b.appendBytes(checkType(byte[].class, v));
+            case PERIOD:
+                return (b, v) -> b.appendPeriod(checkType(Period.class, v));
+            case DURATION:
+                return (b, v) -> b.appendDuration(checkType(Duration.class, v));
+            default:
+                throw new IllegalArgumentException("Unsupported type: " + type);
+        }
+    }
+
+    private static <T> T checkType(Class<T> cl, Object o) {
+        if (o.getClass().isAssignableFrom(cl)) {
+            return (T) o;
+        }
+
+        throw new IllegalArgumentException("Unsupported type: " + o.getClass());
+    }
```

Comment:
See `ClientBinaryTupleUtils.appendValue` - we catch `ClassCastException` and rethrow with more details, this is less code and more efficient than calling `checkType` every time.